azure/Guidelines.md

```diff
@@ -88,80 +86,53 @@ All Azure APIs **MUST** use explicit versioning. The Microsoft REST API guidelin
-
-Even though we recommend clients ignore new fields, there are many libraries and clients that fail when new fields are introduced. Azure services **MUST** update the version number of their API even when adding optional fields. In fact, servers should be as strict as possible. Ignoring a field can result in the API accepting content that containered a typo or an element at the wrong level of nesting. If this missing field changes the semantics (for example, we have seen cases where security settings were misplaced and ignored, leaving the resources more exposed than intended) this can be a huge and hard to discover error.
-
-At a high level, changes to the contract of an API constitute a breaking change. Changes that impact backwards compatibility of an API is also considered a breaking change. Anything that would violate the Principle of Least Astonishment is considered a breaking change in Azure. Below are some concrete examples of what constitutes a breaking change. In the below breaking change scenarios, the API version must be changed.
-
-#### Existing property is removed
-
-If a property called `foo` that was present in v1 of the API needs to be removed, it must be done in a newer API version.
-
-#### New property added to response
-
-If a new property/field is added to the response of an API, the GET-PUT pipeline will be broken. Consider the case where a customer updates the value of a new property "A" from the Azure portal. Another customer does a GET of this resource using the SDK. The SDK will ignore the property since it does not understand it. From the SDK, the customer does a PUT using the model that was returned from the GET. This will overwrite the change made by the first customer from the portal.
-
-#### New required property added to request
+### API Changes that require a version change
 
-If a new property is made required in the request body, clients will have no way to set this and the request will fail.
+Azure services **MUST** update the version number of their API whenever there is a change to the API, no matter how small.  Customers will "lock the API version" so that their code does not fail when the service introduces new features.  They rely on the fact that an API version is a contract with the services that will never change.
 
-#### Property name has changed
+A breaking change is any change in the API that may cause client or service code making the API call to fail. Obvious examples of such a change are the removal of an endpoint, adding or removing a required field or changing the format of the body (from XML to JSON for example). Even though we recommend clients ignore new fields, there are many libraries and clients that fail when new fields are introduced.
```

Comment:
When reviewing ARM APIs, we really have two concepts of a 'breaking change', and differentiating between the two can be quite confusing:
1. Client breaks in an obvious way. This guide covers that case well.
2. A change is introduced in a way that neither old or new client breaks, but it becomes possible for a PUT with a newer version of the client to push some additional state to the server, which will be silently reverted if an older version of the client does a GET + PUT.

Since we have expectations of idempotency in ARM, even the introduction of optional request properties is technically a potentially 'breaking change' as per (2) - although it is tends to be subjective for practical reasons. I know that this guide isn't ARM-specific - just wonder if it's worth explicitly calling out that this scenario isn't considered a breaking change as per these guidelines, but some services may consider as such?

Reply:
Here's part of the problem - we need to get Azure off the ultra-dated concept of PUT as a typical change mechanism.  The industry has moved on and PATCH is the accepted standard for updates now.  This problem of downlevel reversion simply goes away.   How to we get Azure APIs to evolve away from the typical use of PUT for updates?